src/main/java/org/terracotta/build/services/BuildInfo.java

```diff
@@ -0,0 +1,148 @@
+package org.terracotta.build.services;
+
+import org.gradle.api.Project;
+import org.gradle.api.provider.Property;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+
+import java.io.Serializable;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+import static java.lang.Integer.parseInt;
+import static java.util.Arrays.asList;
+import static java.util.stream.Collectors.joining;
+
+public abstract class BuildInfo implements BuildService<BuildInfo.Parameters> {
+  public interface Parameters extends BuildServiceParameters {
+    Property<Git> getGit();
+    Property<Object> getDefaultProjectVersion();
+    Property<Object> getOverrideProjectVersion();
+    Property<Object> getProjectVersion();
+  }
+
+  private final Instant buildTimestamp;
+  private final String branch;
+  private final String commitHash;
+  private final boolean hasLocalChange;
+  private final String revision;
+  private final StructuredVersion version;
+
+  public BuildInfo() {
+    final Git git = getParameters().getGit().get();
+    final String defaultVersion = getParameters().getDefaultProjectVersion().map(Objects::toString).getOrNull();
+    final String overrideVersion = getParameters().getOverrideProjectVersion().map(Objects::toString).getOrNull();
+    final String projectVersion = getParameters().getProjectVersion().map(Objects::toString).get();
+
+    this.buildTimestamp = Instant.now();
+    this.branch = git.getBranch();
+    this.commitHash = git.getCommitHash();
+    this.hasLocalChange = git.hasLocalChange();
+
+    this.version = StructuredVersion.parse(overrideVersion != null ?
+        overrideVersion :
+        (defaultVersion != null ? defaultVersion : projectVersion));
+
+    this.revision = !this.hasLocalChange ?
+        this.commitHash :
+        (this.commitHash + "+" + System.getProperty("user.name") + ":" + git.hash(git.diff(this.commitHash)));
```

Comment:
I'm thinking we might want to nail down some of the presentation aspects of `git diff` here to try and standardize things?

Reply:
what do you mean ? I don't understand... You mean the different options git diff has to generate a diff ?